### PR TITLE
Convert Notification activity and dialog to kotlin.

### PR DESCRIPTION
**Pull request by dbrant**

### CODE REVIEW

File: app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt [hunk 1]
```diff
@@ -0,0 +1,643 @@
...
+
+        // if we're not checking for unread notifications, then short-circuit straight to fetching them.
+        if (displayArchived) {
+            orContinueNotifications
+            return
+        }
+        disposables.add(ServiceFactory.get(WikiSite(Service.COMMONS_URL)).unreadNotificationWikis
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe({ response: MwQueryResponse ->
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt:171]: Please remove the explicit type specification

File: app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+                    val wikiMap = response.query()!!.unreadNotificationWikis()
+                    dbNameMap.clear()
+                    for (key in wikiMap!!.keys) {
+                        if (wikiMap[key]!!.source != null) {
+                            dbNameMap[key] = WikiSite(wikiMap[key]!!.source!!.base)
+                        }
+                    }
+                    orContinueNotifications
+                }) { t: Throwable -> setErrorState(t) })
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt:180]: Please remove the explicit type specification

File: app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+    }
+
+    private val orContinueNotifications: Unit
+        get() {
+            binding.notificationsProgressBar.visibility = View.VISIBLE
+            disposables.add(ServiceFactory.get(WikiSite(Service.COMMONS_URL)).getAllNotifications("*", if (displayArchived) "read" else "!read", currentContinueStr)
+                    .subscribeOn(Schedulers.io())
+                    .observeOn(AndroidSchedulers.mainThread())
+                    .subscribe({ response: MwQueryResponse ->
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt:189]: Please remove the explicit type specification

File: app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+                        onNotificationsComplete(response.query()!!.notifications()!!.list()!!, !currentContinueStr.isNullOrEmpty())
+                        currentContinueStr = response.query()!!.notifications()!!.getContinue()
+                    }) { t: Throwable -> setErrorState(t) })
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt:192]: Please remove the explicit type specification

File: app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+    private val selectedItemCount: Int
+        get() {
+            var selectedCount = 0
+            for (item in notificationContainerList) {
+                if (item.selected) {
+                    selectedCount++
+                }
+            }
+            return selectedCount
+        }
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt:335]: Can this be simplified to the following? 

```
private val selectedItemCount get() = notificationContainerList.count { it.selected }
```

File: app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt [hunk 6]
```diff
@@ -0,0 +1,643 @@
...
+            binding.notificationsRecyclerView.adapter = NotificationItemAdapter()
+        }
+        for (n in notifications) {
+            var exists = false
+            for (nExisting in notificationList) {
+                if (nExisting.id() == n.id()) {
+                    exists = true
+                    break
+                }
+            }
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt:223]: Can this be simplified to the following?

```
notificationList.any { it.id() == n.id() }
```

File: app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt [hunk 7]
```diff
@@ -0,0 +1,643 @@
...
+        if (notificationContainerList.isEmpty()) {
+            binding.notificationsEmptyContainer.visibility = View.VISIBLE
+            binding.notificationsViewArchivedButton.visibility = if (displayArchived) View.GONE else View.VISIBLE
+        } else {
+            binding.notificationsEmptyContainer.visibility = View.GONE
+        }
+    }
+
+    private fun deleteItems(items: List<NotificationListItemContainer>, markUnread: Boolean) {
+        val notificationsPerWiki: MutableMap<WikiSite, MutableList<Notification>> = HashMap()
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt:270]: Please remove the explicit type specification

File: app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt [hunk 8]
```diff
@@ -0,0 +1,643 @@
...
+    }
+
+    private inner class SearchCallback : SearchActionModeCallback() {
+        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
+            actionMode = mode
+            return super.onCreateActionMode(mode, menu)
+        }
+
+        override fun onQueryChange(s: String) {
+            currentSearchQuery = s.trim { it <= ' ' }
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt:563]: Can chnage to `s.trim()`.

File: app/src/main/java/org/wikipedia/notifications/NotificationItemActionsDialog.kt [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+        return binding.root
+    }
+
+    override fun onDestroyView() {
+        _binding = null
+        super.onDestroyView()
+    }
+
+    private fun setUpViewForLink(containerView: View?, iconView: AppCompatImageView?, labelView: TextView?, link: Notification.Link) {
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/notifications/NotificationItemActionsDialog.kt:92]: Any reason why the views are nullable?

**Review by cooltey — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt (hunk 5 above)
Comment (dbrant) [reply app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt:335]: correct!

File: app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt (hunk 6 above)
Comment (dbrant) [reply app/src/main/java/org/wikipedia/notifications/NotificationActivity.kt:223]: yep! but probably better to use `none {}` instead of `any {}` here.

**Pull request merged**